### ISSUE TITLE
Saving files with gzip compression

### ISSUE THREAD

**Issue by Kayne** [theophilusx/ssh2-sftp-client#404]
Hi :)

Using: Node.js 12 (AWS Lambda) and 8.0.0 version of `ssh2-sftp-client`.

I'm creating the postgresql dump using pg_dump, then I wanted to compress with gzip and then send over ssh. This is chunk of my code:

```javascript
// pgdump from https://github.com/jameshy/pgdump-aws-lambda

function pgdump(config, pgDumpSpawnFn = spawnPgDump) {
    return new Promise((resolve, reject) => {
        let headerChecked = false
        let stderr = ''

        // spawn pg_dump process
        const args = buildArgs(config)
        const env = { ...config, LD_LIBRARY_PATH: config.PGDUMP_PATH }
        const process = pgDumpSpawnFn(config.PGDUMP_PATH, args, env)

        // hook into the process
        process.stderr.on('data', data => {
            stderr += data.toString('utf8')
        })

        process.on('close', code => {
            // reject our promise if pg_dump had a non-zero exit
            if (code !== 0) {
                return reject(
                    new Error('pg_dump process failed: ' + stderr)
                )
            }
            // check that pgdump actually gave us some data
            if (!headerChecked) {
                return reject(
                    new Error('pg_dump gave us an unexpected response')
                )
            }
            return null
        })

        // watch the pg_dump stdout stream so we can check it's valid
        const transformer = new Transform({
            transform(chunk, enc, callback) {
                this.push(chunk)
                // if stdout begins with 'PGDMP' then the backup has begun
                // otherwise, we abort
                if (!headerChecked) {
                    headerChecked = true
                    if (chunk.toString('utf8').startsWith('--')) {
                        resolve(transformer)
                    }
                    else {
                        reject(
                            new Error('pg_dump gave us an unexpected response')
                        )
                    }
                }
                callback()
            }
        })

        // pipe pg_dump to transformer
        process.stdout.pipe(transformer)
    })
}
module.exports = pgdump


// uploading
let stream = await pgdump(config)
await sftp.connect({
            host: HOST,
            port: PORT,
            username: USERNAME,
            privateKey: PRIVATE_KEY
        }).then(() => {
            return sftp.put(
                zlib.createGzip( stream ),
                'dump.sql.gz'
            );
        }).then(() => {
            return sftp.end();
        })
        .catch(err => {
            console.error(err.message);
        });
```

I was also trying to do compression in pg_dump function:

```javascript
process.stdout.pipe(transformer).pipe(zlib.createGzip())
```
and just use `stream` in `sftp.put` but with no avail. Every time I got timeout, the ~300 mb dump is not compressed and sent other. There is however file created over SSH but it's size is 0. So looks like it starts the whole process but then just idles untill timeout.

The timeout is after 15 minutes and I pretty sure, the compression should not take so much time on 300mb text file.

Any ideas what am I doing wrong? Clearly I do not understand something obvious. Could you point me in the right direction?

**Comment by theophilusx** [comment]
What version of node?
What version of ssh2-sftp-client?
What platform are you running on?

My guess is that data isn't actually being sent along your pipeline. I would start by confirming that you can create a local gzipped version of the dump. If that fails, I'd step back and verify you can create a plain unzipped local copy of the dump and if that doesn't work, see if you can create a local unzpped dump without the transform stream. 

If you can create a local gzipped version of the dump, I would then try just opening a read stream on that file and using it in the call to put(). If that works, then you at least know something in your pipeline is blocking. It could be that one of the 'nodes' in the pipeline has been paused. I find the order in which you create the pipeline can affect the outcome. I've also found dealing with streams on spawned processes where you are calling a shell command can be very fragile. In fact, the whole reason I got involved with ssh2-sftp-client is because previously I was using a similar approach with lftp as the underlying command rather than pg_dump. While I got it to work, it was terribly fragile. 

If I get time over this long weekend, I'll try to play around with streams and gzipping data. I have done this a few times, but it has been a while. Most of my current projects are only doing the revers i.e. donwloading a zipped file, uncompressing 'on the fly' and writing to the local fs.  Not sure I've done the on the fly compress in an upload since the new ssh2 version was released, so probably good to confirm.

**Comment by Kayne** [comment]
> What version of node?

12

> What version of ssh2-sftp-client?

8.0.0

> What platform are you running on?

AWS Lambda

All of it I answered in the very begining of my post :)

So I did a little testing and with very simple function:

```javascript
exports.handler = async (event) => {
  
// Create file
fs.writeFile('input.txt', 'abc', function (err,data) {
  if (err) {
    return console.log(err);
  }
  console.log(data);
});

let fileContents = fs.readFileSync('input.txt').toString()
console.log(fileContents);
}
```

I got `ENOENT: no such file or directory, open 'input.txt'` on line when I try to read file. So looks like creating files locally isn't working properly, so I didn't bother with further tests.

However, when I just SSH stream over (without gzip compression) the file is successfully transfered! So streams are working properly. It just go weird when I try to pipe through `zlib.createGzip()`.

**Comment by theophilusx** [comment]
Sorry, I didn't notice the platform info at the start. I have had other people finding 'odd' issues running on AWS Lambda. The all seemed to have found solutions, but it did require some workarounds (sorry, don't have any details, but might be worth checking on a lambda specific forum?).

At least it sounds like you have narrowed it down to something relating to zlib. Might be worth experimenting with a test file and zlib on a local system, reading the data via a stream, passing to zlib and writing to a writefile stream and build up form there to passing the stream to ssh2-sftp-stream put() to 'localhost' or similar. At least if you can get it working there and still have issues when running the same under lambda, you will know it is a lambda specific issue. Not a solution, but might help narrow the search space a little. 

One 'off the cuff' though I just had is to be careful regarding encoding. I have seen 'garbage' results because the encoding in a pipeline has gotten screwed up. The 'gzipped' stream should be treated as binary not utf-8 (the default for ssh2-sftp-client's writestream is null i.e. binary). Very slim possibility that if your specifying utf-8 in the initial stram from your spawn process that it gets passed along the pipeline and *might* be causing issues. Around here, we call this a "wild arse guess"!, but .....

**Comment by theophilusx** [comment]
I haven't had time to try out your use case specifically. However, I have made some changes to be released as v9.0.0 which may have baring on your issues. 

The API for `get()` and `put()` methods has changed. In particular, the ability to set autoClose to false on get/put and `end` to false for `pipe()` have been removed. There were problems with these options and client code often set them incorrectly, plus I found what I believe is a bug in ssh2 relating to writeStreams. 

To support use cases where client code needs/wants to set autoClose or end, I have added to new methods - `createWriteStream()` and `createReadStream()`. These are low level methods which return a write or read stream to a remote file on the sftp server. This is all these function do. They do not attempt to manage the stream objects in any way. It is totally down to client code to manage these streams i.e. set any necessary listeners, close and release file descriptors etc. 

While you should still be able to compress/decompress data on the fly with `get()` and `put()`, you should also be able to do the same by using `createReadStream()` and `createWriteStream()`. The big idfference being that with the latter, your responsible for managing and disposing of the stream objects. See the createReadStream and createWriteStream documentation in the ssh2 repository for full details of available options https://github.com/mscdex/ssh2/blob/master/SFTP.md

**Comment by theophilusx** [comment]
I finally got around to trying out 'on the wire' compression using Gzip. I was able to get it working with a standard read stream - have not tried with a postgres stream, but provided it is a standard stream, it should work. 

However, I did notice in your example code you have an error. You have 

```
}).then(() => {
            return sftp.put(
                zlib.createGzip( stream ),
                'dump.sql.gz'
            );

```

that is not the correct syntax for creating a zlib gzip stream object. You need something like 

````
    const gzip = zlib.crteateGzip();
    return sftp.put(stream.pipe(gzip), remotePath);
````

where stream is the source i.e. a read stream, possibly your postgres stream and remotePath is where you want the data put. The stream object will pipe the data to the gzip object which will return a stream which sftp.put will consume to write the file to the remote path. 

As this appears to be a user error and there are no reproducible problems using either a gzip or gunzip stream with the module, I am closing this issue.
